src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+                // and an initializer is presen, meaning we report a diagnostic.ssssssssssssssss
+                context.ReportDiagnostic(field.GetLocation().CreateDiagnostic(RuleExceptionShouldNotInitializeInDeclaration));
+            }
+
+            return;
+        }
+
+        IList<ExpressionStatementSyntax> fieldInitializersInConstructors = fieldInfo.FieldInitializersInConstructors;
+
+        // If the field is initialized in the declaration and in the constructors,
+        // we check if the initializations are diverging.
+        // Since we already checked if the field should not initialize in the declaration,
+        // we know all initializations in the constructors are the same.
+        // So we only need to check if the initializer is diverging from the constructor initializations.
+        if (isInitializerPresent
+            && fieldInitializersInConstructors.Count != 0
+            && HasDivergingInitializations(fieldInitializersInConstructors[0], initializer))
+        {
+            context.ReportDiagnostic(field.GetLocation().CreateDiagnostic(RuleExceptionShouldNotInitializeInDeclaration));
+            return;
+        }
+
+        // We know the initializers in the constructors are the same, so we report diagnostics (general rule) on them.
+        ReportDiagnosticsForInitializersInConstructors(context, fieldInitializersInConstructors);
+    }
+
+    /// <summary>
+    /// Checks if the initializers in the constructors are diverging from the initializer in the declaration.
+    /// </summary>
+    /// <param name="initializerInConstructor">A <see cref="ExpressionStatementSyntax"/> from the constructor.</param>
+    /// <param name="initializerInDeclaration">The initializer from the declaration.</param>
+    /// <returns>A bool on whether the constructor initializers diverge from the declaration.</returns>
+    private static bool HasDivergingInitializations(ExpressionStatementSyntax initializerInConstructor, EqualsValueClauseSyntax initializerInDeclaration)
+    {
+        return !((AssignmentExpressionSyntax)initializerInConstructor.Expression).Right.IsEquivalentTo(initializerInDeclaration.Value);
+    }
+
+    /// <summary>
+    /// Reports diagnostics for initializers in constructors <see cref="GeneralRule"/>.
+    /// </summary>
+    /// <param name="context">The analysis context.</param>
+    /// <param name="initializers">The list of constructor field initializers.</param>
+    private static void ReportDiagnosticsForInitializersInConstructors(SyntaxNodeAnalysisContext context, IList<ExpressionStatementSyntax> initializers)
+    {
+        for (int i = 0; i < initializers.Count; i++)
+        {
+            context.ReportDiagnostic(initializers[i].GetLocation().CreateDiagnostic(GeneralRule));
+        }
+    }
+
+    /// <summary>
+    /// A record to track field initialization information in constructors.
+    /// </summary>
+    private sealed record FieldInitializationInfo
```

Comment:
Why is this a `record`? A `record` is great for data transfers and when you're doing equals operations or don't want to recycle hashing, etc. It's a value in a hash map. Also, comparison is likely not doing what you want comparing two sets of `List<T>`, so if you were doing comparisons you'd want to have your own comparison logic (e.g., the field names were the same based on a string comparer implementation and the lists were of the same length and then have the same contents, etc.)

Reply:
No need for it to be a record. I am changing it to a class.

As for comparison, I don't think I have a need to compare two instances since there should only be one per field. If that becomes something I need then I'll write a comparator